Sources/NIOSOCKS/State/ClientStateMachine.swift

```diff
@@ -0,0 +1,99 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+fileprivate enum ClientState: Hashable {
+    case ready
+    case waitingForAuthenticationMethod(ClientGreeting)
+    case waitingForClientRequest
+    case waitingForServerResponse(ClientRequest)
+    case active
+}
+
+enum ClientAction: Hashable {
+    case none
+    case sendRequest
+    case proxyEstablished
+}
+
+public struct InvalidAuthenticationSelection: Error {
+    
+}
+
+public struct ConnectionFailed: Error, Hashable {
+    public var reply: Reply
+}
+
+struct ClientStateMachine {
+
+    private var state: ClientState
+    
+    var proxyEstablished: Bool {
+        switch self.state {
+        case .active:
+            return true
+        case .ready, .waitingForAuthenticationMethod, .waitingForClientRequest, .waitingForServerResponse:
+            return false
+        }
+    }
+    
+    init() {
+        self.state = .ready
+    }
+
+    mutating func sendClientGreeting(_ greeting: ClientGreeting) {
+        assert(self.state == .ready)
+        self.state = .waitingForAuthenticationMethod(greeting)
+    }
```

Comment:
In the other State Machines I have seen so far the pattern looked a little different:

```swift
    mutating func connectionEstablished() -> Action {
        self.state = .waitingForAuthenticationMethod
        return .sendClientGreeting
    }
```

Reply:
If I were to add a "connection established", it would be in addition to `sendClientGreeting` IMO. @Lukasa do you have an opinion?

Reply:
I think having a state machine edge for this is reasonable enough.